statsmodels/doe/latin.py

```diff
@@ -0,0 +1,198 @@
+
+    for i in range(dim):
+        # Enforce a unique point per grid
+        temp = [np.random.uniform(low=j * step, high=(j + 1) * step)
+                for j in range(n_sample)]
+        np.random.shuffle(temp)
+
+        sample.append(temp)
+
+    sample = np.array(sample).T
+
+    # Sample scaling from unit hypercube to feature range
+    if bounds is not None:
+        min_ = bounds.min(axis=0)
+        max_ = bounds.max(axis=0)
+        sample = sample * (max_ - min_) + min_
+
+    return sample
+
+
+def latin_hypercube(dim, n_sample, bounds=None, centered=False):
+    """Latin hypercube sampling (LHS).
+
+    The parameter space is subdivided into an orthogonal grid of n_sample per
+    dimension. Within this multi-dimensional grid, n_sample are selected by
+    ensuring there is only one sample per row and column.
+
+    Parameters
+    ----------
+    dim : int
+        Dimension of the parameter space.
+    n_sample : int
+        Number of samples to generate in the parametr space.
+    bounds : tuple or array_like ([min, k_vars], [max, k_vars])
+        Desired range of transformed data. The transformation apply the bounds
+        on the sample and not the theoretical space, unit cube. Thus min and
+        max values of the sample will coincide with the bounds.
+
+    Returns
+    -------
+    sample : array_like (n_samples, k_vars)
+        Latin hypercube Sampling.
+
+    References
+    ----------
+    [1] Mckay et al., "A Comparison of Three Methods for Selecting Values of
+    Input Variables in the Analysis of Output from a Computer Code",
+    Technometrics, 1979.
+
+    """
+    if centered:
+        r = 0.5
+    else:
+        r = np.random.random_sample((n_sample, dim))
+
+    q = np.random.random_integers(low=1, high=n_sample, size=(n_sample, dim))
+
+    sample = 1. / n_sample * (q - r)
+
+    # Sample scaling from unit hypercube to feature range
+    if bounds is not None:
+        min_ = bounds.min(axis=0)
+        max_ = bounds.max(axis=0)
+        sample = sample * (max_ - min_) + min_
+
+    return sample
+
+
+def optimal_design(dim, n_sample, bounds=None, doe=None, niter=1, force=False):
+    """Optimal design.
+
+    Optimize the design by doing random permutations to lower the centered
+    discrepancy.
+
+    Centered discrepancy based design show better space filling robustness
+    toward 2D and 3D subprojections. Distance based design better space filling
+    but less robust to subprojections.
+
+    Parameters
+    ----------
+    dim : int
+        Dimension of the parameter space.
+    n_sample : int
+        Number of samples to generate in the parametr space.
+    bounds : tuple or array_like ([min, k_vars], [max, k_vars])
+        Desired range of transformed data. The transformation apply the bounds
+        on the sample and not the theoretical space, unit cube. Thus min and
+        max values of the sample will coincide with the bounds.
+    doe : array_like (n_samples, k_vars)
```

Comment:
IMO `doe` is not a good name here, not specific enough and same as folder.
maybe `start_design`

a shortcut name can be used as alias within the function